src/rebar_git_resource.erl

```diff
@@ -146,11 +158,10 @@ download(Dir, AppInfo, State) ->
     download_(Dir, AppInfo, State).
 
 download_(Dir, {git, Url}, State) ->
-    ?WARN("WARNING: It is recommended to use {branch, Name}, {tag, Tag} or {ref, Ref}, otherwise updating the dep may not work as expected.", []),
-    download_(Dir, {git, Url, {branch, "master"}}, State);
+    ?DEBUG("Git revision is not specified, using origin/HEAD", []),
+    download_(Dir, {git, Url, {ref, "origin/HEAD"}}, State);
 download_(Dir, {git, Url, ""}, State) ->
-    ?WARN("WARNING: It is recommended to use {branch, Name}, {tag, Tag} or {ref, Ref}, otherwise updating the dep may not work as expected.", []),
```

Comment:
I'd keep the warnings. It is probably a good idea to make dep updates readable/predictable for people looking at the source config, regardless of the lock files.

Reply:
Oh yea, definitely should keep that warning.

Reply:
But in that case, what shall be the "clean" (no-warning) way to depend on the default branch for a Git repo? To specify {ref, "origin/HEAD"} or "origin/HEAD" (old-rebar-compatible way) explicitly? The latter also generates a warning. In addition, both ways definitely do not upgrade as expected: `needs_update()` always returns `true` for any symbolic ref (HEAD, origin/HEAD, etc), because it doesn't bother to resolve the symbolic ref to the actual commit hash that it has in the lock file, and ends up comparing apples to oranges. So `rebar3 upgrade` may fail or rewrite local uncommitted changes even when there are no changes upstream. (Shall I fix that in a separate commit?) On the contrary, omitting the ref altogether now works as expected (more or less) wrt `rebar3 upgrade`: it will only try to check out when there are some changes upstream. So IMO the warning (the way it was originally phrased) is no longer relevant.

Reply:
The clean way would be to check what the default branch is. There may be a few different used most often, I know I've seen at least `develop`, `master`, `main`, but actually being changed is almost never?

Reply:
But I've seen repositories where the default branch has changed (old branch was no longer present), which broke our builds. The main intent of this PR is to cleanly avoid breakage in such cases, and that is to point a Git dependency to the default upstream branch, whatever it is at the moment. I understand that deleting or renaming default branches is largely the fault/bad practice on that part of the repository owner (if you change your default branch, you should at least keep the old branch for compatibility) -- but that is, of course, outside of our control.

Reply:
And I spoke too soon -- in the new default branch case, there still remained 2 scenarios not handled consistently with the usual `{branch, Name}` case, namely:
1) When there is no `rebar.lock` and there are locally committed changes which are not present upstream, I believe the expectation is that `rebar3 upgrade` should *not* reset the HEAD to upstream (albeit there are neither test cases nor documentation for this, so I cannot be sure what the expectation is).
2) When the `origin` URL has changed, upgrade should always be performed.
This I fixed, so I think now the default branch case would be handled consistently with {branch, Name}.

Reply:
I don't necessarily worry about 1, since we can't easily support people changing deps in the _build directory with no other indications and have never planned to.

As for 2, yeah that should more or less always happen, and I believe the update comparison steps do compare URLs for that to happen.

That being said, I'd still want the warning to be in place. It is a reasonable one to keep around mostly because we also can't provide any guarantees that the rebar3 copy you're using is as up-to-date as the copy other users may be using and for the foreseeable future, we have to account for the fact that various teams or groups of contributors are going to be working with mixed versions where the usage of a default branch will not be consistent across everyone.

As such, even if the next rebar3 release is safely using `origin/HEAD`, you may still end up with contributors who don't and for whom the default upgrade mechanism will default to `master` regardless of having switched to `main` with both branch available, and will yield surprising upgrade or fetching behaviours. In a wider ecosystem, it remains true that `{branch, Name}, {tag, Tag} or {ref, Ref}` is safer than specifying nothing.